src/tools/rbd_mirror/image_replayer/PrepareLocalImageRequest.cc

```diff
@@ -124,9 +126,25 @@ void PrepareLocalImageRequest<I>::handle_get_mirror_info(int r) {
     return;
   }
 
-  // TODO save current mirror state to determine if we should
-  // delete a partially formed image
-  // (e.g. MIRROR_IMAGE_STATE_CREATING/DELETING)
+  if (m_mirror_image.state == cls::rbd::MIRROR_IMAGE_STATE_CREATING) {
+    dout(5) << "local image is still in creating state, issuing a removal"
+            << dendl;
+    auto ctx = new LambdaContext([this](int) {
+        finish(-ENOENT);
+    });
+    ImageDeleter<I>::trash_move(m_io_ctx, m_global_image_id,
+                                false, m_threads->work_queue, ctx);
+    return;
+  } else if (m_mirror_image.state == cls::rbd::MIRROR_IMAGE_STATE_DISABLING) {
+    dout(5) << "local image mirroring is being disabled, retrying in 1 minute"
+            << dendl;
```

Comment:
I am not sure we need to wait when the state is disabling. It is means that the image is already moved to trash and it should not "disturb" the process of the new image creation. So it seems to me it is correct to just return `-ENOENT` here, as it was in your previous version.

Note, for the "creating" case, after "trash_move" is returned you will also have the state "disabling".

There is a possibility though, that the rbd-mirror is interrupted in  "trash_move",  just after it set the state to disabling but before moving to trash. Then after the rbd-mirror restart it will "think" the image is already in the trash when it is not. We probably could improve this but we can leave it out of scope of this PR.

Reply:
Ok I reverted it!